### PR TITLE
Convert xds bootstrap code to use new JSON API

### PR DESCRIPTION
**Pull request by markdroth**

### CODE REVIEW

File: src/core/ext/filters/client_channel/xds/xds_bootstrap.cc [hunk 1]
```diff
@@ -246,242 +187,81 @@ grpc_error* XdsBootstrap::ParseChannelCreds(grpc_json* json, size_t idx,
...
-      InlinedVector<grpc_error*, 1> parse_errors =
-          ParseMetadataStruct(child, &node_->metadata);
-      if (!parse_errors.empty()) {
-        grpc_error* parse_error = GRPC_ERROR_CREATE_FROM_VECTOR(
-            "errors parsing \"metadata\" object", &parse_errors);
-        error_list.push_back(parse_error);
-      }
+  auto it = json->mutable_object()->find("id");
+  if (it != json->mutable_object()->end()) {
+    if (it->second.type() != Json::Type::STRING) {
```
Comment (muxi) [inline src/core/ext/filters/client_channel/xds/xds_bootstrap.cc]: Confirm `id` field is optional?

File: src/core/ext/filters/client_channel/xds/xds_bootstrap.h [hunk 2]
```diff
@@ -19,6 +19,8 @@
 
 #include <grpc/support/port_platform.h>
 
+#include <memory>
```
Comment (muxi) [inline src/core/ext/filters/client_channel/xds/xds_bootstrap.h]: What is `memory` for?

File: src/core/ext/filters/client_channel/xds/xds_api.cc [hunk 3]
```diff
@@ -105,56 +106,59 @@ bool XdsDropConfig::ShouldDrop(
 namespace {
 
 void PopulateMetadataValue(upb_arena* arena, google_protobuf_Value* value_pb,
-                           const XdsBootstrap::MetadataValue& value);
+                           const Json& value);
```
[the remark this answers is not in the document]
Comment (markdroth) [reply src/core/ext/filters/client_channel/xds/xds_api.cc]: I don't see where `PopulateMetadataValue()` is declared outside of the anonymous namespace.  It looks to me like it's declared and defined in the anonymous namespace.

Putting a function in an anonymous namespace is the same thing as declaring it `static`, but it's a more idiomatic C++ way of doing it.  (`static` is the old C style.)

File: src/core/ext/filters/client_channel/xds/xds_bootstrap.h (hunk 2 above)
Comment (markdroth) [reply src/core/ext/filters/client_channel/xds/xds_bootstrap.h]: It's for `std::unique_ptr<>`.  This should have been here before this PR anyway, so I fixed it while I was in here.

File: src/core/ext/filters/client_channel/xds/xds_bootstrap.cc (hunk 1 above)
Comment (markdroth) [reply src/core/ext/filters/client_channel/xds/xds_bootstrap.cc]: Yeah, I don't see any reason it should be required.  Even if the server wants to require it, the client doesn't need to impose that requirement.

File: src/core/ext/filters/client_channel/xds/xds_api.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (muxi) [reply src/core/ext/filters/client_channel/xds/xds_api.cc]: Sorry my bad... Somehow I mismatched the braces.

File: src/core/ext/filters/client_channel/xds/xds_bootstrap.cc (hunk 1 above)
Comment (muxi) [reply src/core/ext/filters/client_channel/xds/xds_bootstrap.cc]: Hmm ok. It was just a bit weird to see that a node (or anything) having an optional "identifier" field. But I guess it's possible if the picking algorithm just doesn't care which node it is.

**Comment by markdroth** [comment]
Known issues: #19627 #19966 #21844 #16201 #21230 #20198

File: src/core/ext/filters/client_channel/xds/xds_bootstrap.cc [hunk 4]
```diff
@@ -39,129 +39,84 @@ std::unique_ptr<XdsBootstrap> XdsBootstrap::ReadFromFile(grpc_error** error) {
...
+  for (Json& child : *json->mutable_array()) {
+    if (child.type() != Json::Type::OBJECT) {
       char* msg;
       gpr_asprintf(&msg, "array element %" PRIuPTR " is not an object", idx);
       error_list.push_back(GRPC_ERROR_CREATE_FROM_COPIED_STRING(msg));
     } else {
-      grpc_error* parse_error = ParseXdsServer(child, idx);
+      grpc_error* parse_error = ParseXdsServer(&child, idx);
       if (parse_error != GRPC_ERROR_NONE) error_list.push_back(parse_error);
     }
```
Comment (muxi) [inline src/core/ext/filters/client_channel/xds/xds_bootstrap.cc]: Should be an `idx++` here?

Comment (markdroth) [reply src/core/ext/filters/client_channel/xds/xds_bootstrap.cc]: Good catch!  Sent you #21866 to fix this.

**Pull request merged**